horovod/spark/common/datamodule.py

```diff
@@ -0,0 +1,79 @@
+import importlib
+from abc import ABC, abstractmethod
+
+
+_data_modules = {}
+
+
+class DataModule(ABC):
+    """Context manager base class for data module/loader implementations."""
+    short_name = None     # implementations should provide a short name for easy reference, e.g. 'petastorm', 'nvtabular', etc.
+
+    def __init__(self, train_dir: str, val_dir: str, num_train_epochs: int=1, has_val: bool=True,
+                 train_batch_size: int=32, val_batch_size: int=32, shuffle_size: int=1000,
+                 transform_fn=None, inmemory_cache_all=False,
+                 cur_shard: int=0, shard_count: int=1, schema_fields=None, storage_options=None,
+                 steps_per_epoch_train: int=1, steps_per_epoch_val: int=1, verbose=True,
+                 debug_data_loader: bool=False, train_async_data_loader_queue_size: int=None,
+                 val_async_data_loader_queue_size: int=None, **kwargs):
+        super().__init__()
+        self.train_dir = train_dir
+        self.val_dir = val_dir
+        self.num_train_epochs = num_train_epochs
+        self.has_val = has_val
+        self.train_batch_size = train_batch_size
+        self.val_batch_size = val_batch_size
+        self.shuffle_size = shuffle_size
+        self.transform_fn = transform_fn
+        self.inmemory_cache_all = inmemory_cache_all
+        self.cur_shard = cur_shard
+        self.shard_count = shard_count
+        self.schema_fields = schema_fields
+        self.storage_options = storage_options
+        self.steps_per_epoch_train = steps_per_epoch_train
+        self.steps_per_epoch_val = steps_per_epoch_val
+        self.verbose = verbose
+        self.debug_data_loader = debug_data_loader
+        self.train_async_data_loader_queue_size = train_async_data_loader_queue_size
+        self.val_async_data_loader_queue_size = val_async_data_loader_queue_size
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        pass
+
+    @abstractmethod
+    def train_data(self, reader=None):
+        """Returns the training data in a form required by the target DL framework."""
+        pass
+
+    @abstractmethod
+    def val_data(self, reader=None):
+        """Returns the validation data in a form required by the target DL framework."""
+        pass
+
+    @classmethod
+    def register(cls):
+        """Adds this DataModule implementation to a global registry keyed by short name."""
+        _data_modules[cls.short_name] = cls
```

Comment:
What if that class data module name exists already in `_data_modules`? This then silently overwrites that class, which would be hard to discover.

This method should check if `cls.short_name` already exists in `_data_modules`, and log a warning if it does. Further, `cls` should not be registered in that case by its short name. It can still be accesses by the fully qualified class name (which should be logged in the warning as well).

Alternative is to also drop the existing class from `_data_modules`, so neither datamodule is accessible through the collided class name and either has to be accessed through the fully qualified datamodule class name.

Reply:
Ended up removing this registry, since it's likely more trouble than it's worth.  Now, user can just import the datamodule class itself and use the class directly.